DOC: Add documentation about how to contribute

jhlegarreta opened the pull request:
Add documentation about how to contribute to the toolkit.

jhlegarreta commented:
Some context about this PR.

I think it is becoming increasingly more important to agree on and stick to a set of principles to keep our code healthier. The code is read many more times than it is written, and keeping a consistent style improves the readability. Thus, abiding by these simple rules would help to decode the statements with less effort.

Using automated tools such as `black` will help us make the code consistent, and hopefully we will no longer need (or drastically reduce) the PEP8 style commits. We may not be able to control how these automated tools format every aspect of the code, or we may not like how some aspects are handled (think of the eternal debate between the `Allman` and `Whitesmiths` styles, for example), but it is a compromise. I personally prefer to read code that is formatted consistently, even if I do not like the how it is formatted. Also, some formatting choices go beyond what may fall into one's preferences and help avoid bugs (e.g. using braces for one-line flow control statements in `C++`).

I know it is not easy to fully agree on every aspect, but IMHO we should be able to agree on a minimum set of things. Using these tools frees us from doing this manually  (and inevitably failing to do it consistently), and being faced to the eternal question or where to break a long lines, for example.

A similar reasoning concerning their readability applies to the commit messages. Writing commit messages that comply with a set of rules allows to convey some important information about the rationale behind the set of changes: e.g. why they are necessary, what feature they bring or what bug they fix. Having clear commit message subjects with some given prefixes also allows us to hopefully find a given change with less effort. Additionally, if at some point we will be willing to generate a shortlog or release notes, they will also summarize better the work done.

Also, I saw the new card about the style in trello. I'd dare to say that having them here helps to track changes and to document the repository better. It will potentially be useful for other people reading the code as well.

So we may want to have a look at the available tools to do the code formatting, edit, discuss and agree on some reasonable grounds for both the code formatting and the commit messages, and try to stick to them.

Thanks for considering this !

AntoineTheb commented:
I agree that having a set of rules defining code style, doc style, etc. is super important. However I think that this is a discussion that needs to be had in a either in person when you come back or atleast over Skype, so we can all clearly agree on what the purpose of this repo is and what shoulds be enforced, tested, documented. Still, great job on the documentation !

jhlegarreta commented:
@AntoineTheb Thanks for considering this. Sounds reasonable. Will be back next Thursday so we can have a first discussion.

EmmaRenauld commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,109 @@
+
+If you are adding code into a module that already has a test file, add
+additional tests into the respective file.
+
```
Just opening a discussion.

How are we going to be able to test anything? Tools in dwi_ml are going to be a base ("abstract"), so that people can create their own classes and scripts for their own data. So things will not really be meant to be fully runnable probably (so full scripts?). I guess at some point, the tests are more gonna come from each of our projects, no? I'm not sure we can force people to add tests.

jhlegarreta replied:
I think there is a non-negligible amount of code exclusive to this repository and that could/will eventually be shared among our projects that can be tested. As right now I am not using this code base, I would need to look at specific examples. But the moment when two projects use the same methods is already a powerful reason to think about tests. Even when a method is used by just one project, having a test allows you to be able to reproduce your results or not to break things inadvertently, and be able to demonstrate a few basic things that could be called into question otherwise.

As long as we do not share this viewpoint as owners of the repository, we are not going to be able to enforce writing tests, meaning that we should not merge something that is not tested, including our own code.

A test allows to self-answer questions, questions from other people, and more importantly, to avoid bugs (bugs that may lead to unexpected and incorrect results).

IMHO, we just need to look at the packages we rely on to realize about the importance of this.

Writing tests takes time, and that is a trade-off between other things that need time and writing tests. But advising to do so explicitly is already a step in the right direction.

EmmaRenauld commented on doc/devel/coding_style_guideline.rst:
```diff
@@ -0,0 +1,169 @@
+=============================
+DWI_ML Coding Style Guideline
+=============================
```
We have already started adding such a page in the scil-documentation. Maybe you could improve this page: https://scil-documentation.readthedocs.io/en/latest/coding/scilpy.html
(you can create a PR in https://github.com/scilus/scil-documentation)

and here we could simply point there.

Actually, thinking about it, it would be nice if you did the same for your github explanation higher. We could just tell people that we ask for the same guidelines as the scil lab.

jhlegarreta replied:
Sorry, but I would just target this repository. I prefer to be cautious about `Scilpy`. Here we have the opportunity to decide on what we want our code and coding practices look like at this early stage.

Also, by the same token, one may argue why we do not adopt the `VITALabAI` style, in case the two encourage or adopt different styles.

Ideally we would be using `Scilpy` and `VITALabAI` as a `pip` installable packages, and the way things are documented there should not interfere they way things are documented here.

IMHO the `SCIL` documentation site is a necessary resource, and an extremely useful one for all lab members, but the coding practices should be written in a code repository's documentation resources, not in the documentation resources of a lab. Think that `SCIL` has other projects, or uses other scripting languages as well.

EmmaRenauld replied:
The page I cited is not on the scilpy page but on the scil-documentation page. It means it could be general tools for all students in our lab. So I think it could still be moved to our doc.

I understand your point about scil vs vital. Do they already have an official coding style?

I just think that the work you are doing is very nice and exactly what I wanted to add in the scil-documentation. I would hate to just copy and paste your work there, and I would hate even more to have to remember to code differently when coding here or in scilpy.

Or at least we could see what's common. And then for the differences, we could deal with them. Ex, here we tend to add the annotation for parameters (ex: sft: StatefulTractogram = none) and not in scilpy. We could maybe explain it here in this smaller doc or then decide to do as in scilpy or then convince scilpy to do it.

Maybe others have an opinion? @mdesco

jhlegarreta replied:
> I understand your point about scil vs vital. Do they already have an official coding style?

There is no specific coding style guideline in either lab beyond an explicit mention to complying with `PEP8` in `VITALabAI`.

> I just think that the work you are doing is very nice and exactly what I wanted to add in the scil-documentation. I would hate to just copy and paste your work there, 

Thanks for this Emmanuel. I appreciate it a lot. As said, I am not willing to open this front at SCIL or VITAL.

> and I would hate even more to have to remember to code differently when coding here or in scilpy.

Different organizations may adopt different conventions. I personally find that if there is a tool that automatically formats the code, then I just do not need to worry too much about it. In my experience, if such a tool exists in a repository, then I do not care about the style as I would do otherwise, and this implies accepting a style that I may dislike if I had to format manually, and getting used to reading code with that style.

> Ex, here we tend to add the annotation for parameters (ex: sft: StatefulTractogram = none) and not in scilpy.

If those of you that have been using it find it useful, then 👍 for adding it to these guidelines and I would try my best to stick to it.,

mdesco commented:
Hi all, 

I just read this discussion and I have mixed feelings about it.  On one hand, giving us strict and rigid rules can be good and, on the other, having the flexibility to code/review/merge "fast" in this research-code context... I have seen both in my career (dipy vs scilpy)... And I see dipy is truly a hurdle for lots of research-developers, new internship students, etc... 

Maybe it would be best for us to have this discussion by voice on a call this week, if possible? I know it is hard for @ppoulin91 to meet but can we try between myself, @jhlegarreta @jodoin @AntoineTheb @EmmaRenauld ?

I suggest @jhlegarreta prepares a minimum set of standards and rules that he would like see implemented and others prepare a mental list or counter-list as well.

Cheers

jhlegarreta commented:
This was added for the purpose of serving as a guideline. No intention to be a burden or a show-stopper.

I am available for a live discussion 👍 !

EmmaRenauld commented:
Available! We can decide a time on slack. After lunch is probably the best for me.

AntoineTheb commented:
Which day @EmmaRenauld ? Should we make a Doodle to arange a meeting ?

EmmaRenauld commented:
@jhlegarreta , can you check how complicated it would be to fix conflicts? If it's too complicated, wait after our meeting!

EmmaRenauld commented:
> former `README` file ([1d2ba98](https://github.com/scil-vital/dwi_ml/commit/1d2ba98d8d9a279eda6573b3a0c1be55a51a5f06) _vs._ [219703b](https://github.com/scil-vital/dwi_ml/commit/219703bc461ece125c7544b83f54cf143f13316a)) had useful informatio

I just didn't want to have the same info in two different places. But if you want to move back the "Getting started" page (https://dwi-ml.readthedocs.io/en/latest/getting_started.html) in the readme.md, go for it.

jhlegarreta commented:
I'd dare to say that adding 
```
.. include:: ../README.rst
```
to `doc/index.rst` or else, adding a `docs/README.rst` symlink to the root `README.rst` and including the former in `doc/index.rst` should do the trick.

EmmaRenauld commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,109 @@
+--------
+
+We use `pytest`_ to write the tests of the code.
+
```
Do we test only script, or all code functions? Do we test classes? Do we test classes that are already tested in a script?  Should there be more details here? Up to you, I am not really used to this.

jhlegarreta replied:
I would leave it as it is. No hard requirements. If people are willing to write them because it helps them (and possibly others), and have some time, then let them write them to the detail they are willing to.

Ideally, we should write tests for both. But our time limitations and other constraints may prevent us from getting to that point. I am fine with that.

If we write one every once in a while, at least we know the testing tool the repository uses.

> Do we test classes that are already tested in a script?

A script as we understand them in the context of `Scilpy`, `VITALabAI` or this repository is not really a test, and does not test the classes used either. A test would ideally compare the result of some call against an baseline that is known to be correct.

EmmaRenauld commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,109 @@
+Code style
+----------
+
+Code contributions should be formatted according to the `DWI_ML Coding Style Guideline <./doc/devel/coding_style_guideline.rst>`_
```
Missing a point at the end of the line. guideline.rst>`_ .

jhlegarreta replied:
Done in ad7ee3c.

EmmaRenauld commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,109 @@
+
+Coverage
+--------
+
```
I would move the two first lines

(We use `pytest`_ to write the tests of the code.) (If you are adding code into a module that already has a test file, add additional tests into the respective file.) before the title "coverage".  And under a title "testing"

jhlegarreta replied:
:+1: Done in ad7ee3c.

EmmaRenauld commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,109 @@
+reviews.
+
+These guidelines are largely inspired by Chris Beam's `How to Write a Commit Message <https://chris.beams.io/posts/git-commit/>`_
+post.
```
I see that it was also taken from ITK. Should it be cited?

jhlegarreta replied:
I was the one that added that to ITK with the contributions of other folks, but yeah, why not.

jhlegarreta replied:
Done in ad7ee3c.

EmmaRenauld commented on doc/devel/coding_style_guideline.rst:
```diff
@@ -0,0 +1,169 @@
+
+The main principles behind DWI_ML development are:
+
+* **Robustness**: the results of a piece of code must be verified
```
Again, should you cite Dipy here for taking the same text? Is it plagiarism or is it fine?

jhlegarreta replied:
I added that to DIPY, probably inspired from many places, but yeah, why not.

jhlegarreta replied:
Done in ad7ee3c.

EmmaRenauld commented on doc/devel/coding_style_guideline.rst:
```diff
@@ -0,0 +1,169 @@
+* Keep name length within a four word maximum, and avoid gratuitous context.
+* Only use correctly-spelled dictionary words and abbreviations that appear in a
+  dictionary. Make exceptions for ``id`` and documented domain-specific
+  abbreviations (e.g. ``fa``, ``md`, ``acc``, ``pos``).
```
Missing a `
``md``,

jhlegarreta replied:
Wow, nice, thanks. Done in ad7ee3c.

EmmaRenauld commented on doc/devel/coding_style_guideline.rst:
```diff
@@ -0,0 +1,169 @@
+Package names
+^^^^^^^^^^^^^
+
+* Use a noun name: e.g. ``learning`` instead of ``learners``. An exception could
```
Not sure I understand the sentence. Don't you mean "learners instead of learning"? And model is a noun, how is it an exception? Or then you meant "use an action name"?

jhlegarreta replied:
> Don't you mean "learners instead of learning"?

Right the opposite: the package name should be e.g. `learning` instead of `learners`, or `filtering` instead of `filters`. I can change it to `action name` to make it clearer.

jhlegarreta replied:
Made it more clear in ad7ee3c.

EmmaRenauld commented on doc/devel/coding_style_guideline.rst:
```diff
@@ -0,0 +1,169 @@
+  negation in boolean names. Do not use names that require a prefix like ``not``
+  that inverts the variable’s truth value.
+
+Documentation
```
Maybe I would move this in a separated page? python coding styyle and rst coding style.

jhlegarreta replied:
For now, I would avoid splitting it, since that would mean that we should have three files, the parent including the two childs (code; rst), and I think it is fine as it is at least for now. Also, the documentation is intimately linked to the code.

jhlegarreta commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,109 @@
+New contributions are required to have as close to 100% code coverage as
+possible. This means that the tests written cause each and every statement in
+the code to be executed, covering corner-cases, error-handling, and logical
+branch points. To check how much coverage the tests have, you will need.
```
Nice catch. Addressd in ad7ee3c. I removed it. The next paragraphs explains how to do it locally.

jhlegarreta commented:
@EmmaRenauld I'd be grateful if you could test [this](https://github.com/scil-vital/dwi_ml/pull/26#issuecomment-610533833). Including it should be enough; if not, the symlink option may work. For a quick test, you can just include the `README` file as it is. You may want to change it to `rst` syntax to ensure that it will work when converting the `README` back to rst in this PR. If that works, I'd rebase on `master` and fix conflicts.

jhlegarreta commented:
Sorry for the delay. I checked that including the original `README.rst` in the `index.rst` through an indirect include file works. Screenshots below.

So I rebased on `master`, got back the original `README.rst`, removed redundant contents (e.g. `License` and `Citation` sections) from the `doc/index.rst`, added the documentation badge and pushed a new commit (had to force push due to the rebase/conflicts).

This is ready on my end. If it gets green light, I can squash the commits, push again and it would be ready to be merged.

![dwi_ml_doc_index](https://user-images.githubusercontent.com/5576557/79281929-d523de00-7e81-11ea-9bf3-68e88dd3b050.png)

![dwi_ml_doc_readme](https://user-images.githubusercontent.com/5576557/79281946-e10fa000-7e81-11ea-886b-edbefbc02cb7.png)

EmmaRenauld commented on doc/index.rst:
```diff
@@ -1,55 +1,12 @@
     :maxdepth: 1
     :caption: Table of contents
 
+    readme_link
```
Tested. This works.

Or maybe you could add the toctree but instead of adding a page readme_link, you could write this part` .. include:: ../README.rst` under the toctree. This way, the main page of the website would already be the same as the main page of the github (except the table of contents), and you wouldn't need the readme_link page anymore.

EmmaRenauld replied:
Weird. I tried it and it worked. If you think it would be better we can skype to find out why.

jhlegarreta replied:
I spent quite some time trying to make that work on my system without success, so if the current solution works for those trying the PR, then I'd go forward with it. You can still open an issue if you feel like you would like to try and find out a why it would not work on all systems/how to make it work.

EmmaRenauld commented:
I also discovered an error from a previous PR. Could you fix it here at the same time?

Line 12 in setup.py:
```
with open(path.join(here, 'index.rst'), encoding='utf-8') as f:
    long_description = f.read()
```
The file index.rst does not exist anymore. Can you change for doc/index.rst? Or for the file where the long_description is? Thanks.

jhlegarreta commented:
> The file index.rst does not exist anymore. Can you change for doc/index.rst? Or for the file where the long_description is? Thanks.

I would say that reading all contents from `doc/index.rst` would just store strings like the `toctree` or the strings that in reality are filenames. That is, there is no such token (`long_description`) in that file, and it has not been present in any file to my knowledge. So to fix things:
- We could create a separate file to host tokens relevant to the library information.
- Add a docstring to the _setup.py_ file, e.g.
```
"""
DWI ML is a toolkit for ...
"""
```
and do, e.g.:
```
DOCLINES = (__doc__ or '').split("\n")
...
description = DOCLINES[0]
long_description = "\n".join(DOCLINES[2:])
```
- Get away with reading the `README.rst` for now, e.g.
```
long_description = Path("README.rst").read_text(encoding="utf-8"),
```

I do not have a preference for either for now.

EmmaRenauld commented:
> > The file index.rst does not exist anymore. Can you change for doc/index.rst? Or for the file where the long_description is? Thanks.
> 
> I would say that reading all contents from `doc/index.rst` would just store strings like the `toctree` or the strings that in reality are filenames. That is, there is no such token (`long_description`) in that file, and it has not been present in any file to my knowledge. So to fix things:
> 
> * We could create a separate file to host tokens relevant to the library information.
> * Add a docstring to the _setup.py_ file, e.g.
> 
> ```
> """
> DWI ML is a toolkit for ...
> """
> ```
> 
> and do, e.g.:
> 
> ```
> DOCLINES = (__doc__ or '').split("\n")
> ...
> description = DOCLINES[0]
> long_description = "\n".join(DOCLINES[2:])
> ```
> 
> * Get away with reading the `README.rst` for now, e.g.
> 
> ```
> long_description = Path("README.rst").read_text(encoding="utf-8"),
> ```
> 
> I do not have a preference for either for now.

Me neither. As you wish.

pep8speaks commented:
Hello @jhlegarreta! Thanks for updating this PR. 

* In the file [`dwi_ml/info.py`](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py):

> [Line 28:12](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L28): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 28:80](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L28): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (120 > 79 characters)
> [Line 30:18](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L30): [F821](https://duckduckgo.com/?q=pep8%20F821) undefined name 'path'
> [Line 30:31](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L30): [F821](https://duckduckgo.com/?q=pep8%20F821) undefined name 'path'
> [Line 32:22](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L32): [F821](https://duckduckgo.com/?q=pep8%20F821) undefined name 'path'
> [Line 39:5](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L39): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 40:11](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L40): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 41:17](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L41): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 42:12](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L42): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 43:17](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L43): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 45:4](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L45): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 46:13](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L46): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 47:12](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L47): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 48:14](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L48): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 49:12](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L49): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 50:8](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L50): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 51:12](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L51): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 52:9](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L52): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 53:7](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L53): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 54:13](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L54): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 55:10](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L55): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 56:6](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L56): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 57:6](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L57): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 58:6](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L58): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 59:10](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L59): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 60:8](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L60): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 61:9](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L61): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 62:9](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/info.py#L62): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator

* In the file [`dwi_ml/setup_helpers.py`](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/setup_helpers.py):

> [Line 5:1](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/setup_helpers.py#L5): [F401](https://duckduckgo.com/?q=pep8%20F401) 'os' imported but unused
> [Line 6:1](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/dwi_ml/setup_helpers.py#L6): [F401](https://duckduckgo.com/?q=pep8%20F401) 'sys' imported but unused

* In the file [`setup.py`](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/setup.py):

> [Line 22:50](https://github.com/scil-vital/dwi_ml/blob/ebccd76e665fcd0319fb2a7262dd36e8137da4e6/setup.py#L22): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','

jhlegarreta commented:
(Very sorry for the PEP8 noise in the previous attempts. Let's hope we can set up something to address those automatically)

> Me neither. As you wish.

Done in d00ad64.

Comment about the hooks added to `CONTRIBUTING.md` in 2964a65.

Note that this changes `setup.py` and hence [this change](https://github.com/scil-vital/dwi_ml/blob/cc8a54f24748a79669e7fc84670df5c6161a912d/setup.py#L12) will cause a conflict. The `index.rst` file contains the ToC which is exclusive to the documentation and does not fit in the eventual package information. So I'd rather remove that edition from PR #37.

We will leave reading the release version from the `info.py` file to [feed this line](https://github.com/scil-vital/dwi_ml/blob/master/doc/conf.py#L25) for another PR. 

All comments addressed. This is ready.

EmmaRenauld commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,129 @@
+----------
+
+Code contributions should be formatted according to the `DWI_ML Coding Style Guideline <./doc/devel/coding_style_guideline.rst>`_.
+Please, read the document to conform your code contributions to the DWI_ML
```
I have a small last comment. Would you mind moving this (`./doc/devel/coding_style_guideline.rst`) out of the "doc" folder?   `./coding_style_guideline.rst`
Right now, when building the html, it warns that this page is not included in any toctree, and I fear I will always be wondering why. I think it could be moved at the same place as this page (contributing).

jhlegarreta replied:
I'd vote for living with that until we find a solution to address that in another PR while keeping it in the current folder. You may want to open an issue as a reminder. Having all doc-related files within one folder avoids the potential clutter of the root folder.

jhlegarreta commented:
> However, installing dwi_ml does not seem to work for me:

Sorry, my bad. Fixed in 7fac5ab.
